cmd/snap/cmd_info.go

```diff
@@ -305,37 +226,217 @@ func printDescr(w io.Writer, descr string, termWidth int) error {
+	arg        string
+	termWidth  int
+	fmtTime    func(time.Time) string
+	absTime    bool
+	verbose    bool
+}
+
+func (iw *infoWriter) maybePrintPrice() {
+	if iw.resInfo == nil {
+		return
+	}
+	price, currency, err := getPrice(iw.remoteSnap.Prices, iw.resInfo.SuggestedCurrency)
+	if err != nil {
 		return
 	}
+	fmt.Fprintf(iw, "price:\t%s\n", formatPrice(price, currency))
+}
 
-	commands := make([]string, 0, len(allApps))
-	for _, app := range allApps {
+func (iw *infoWriter) maybePrintType() {
+	// XXX: using literals here until we reshuffle snap & client properly
+	// (and os->core rename happens, etc)
+	t := iw.theSnap.Type
+	switch t {
+	case "", "app", "application":
+		return
+	case "os":
+		t = "core"
+	}
+
+	fmt.Fprintf(iw, "type:\t%s\n", t)
+}
+
+func (iw *infoWriter) maybePrintID() {
+	if iw.theSnap.ID != "" {
+		fmt.Fprintf(iw, "snap-id:\t%s\n", iw.theSnap.ID)
+	}
+}
+
+func (iw *infoWriter) maybePrintTrackingChannel() {
+	if iw.localSnap == nil {
+		return
+	}
+	if iw.localSnap.TrackingChannel == "" {
+		return
+	}
+	fmt.Fprintf(iw, "tracking:\t%s\n", iw.localSnap.TrackingChannel)
+}
+
+func (iw *infoWriter) maybePrintInstallDate() {
+	if iw.localSnap == nil {
+		return
+	}
+	if iw.localSnap.InstallDate.IsZero() {
+		return
+	}
+	fmt.Fprintf(iw, "refresh-date:\t%s\n", iw.fmtTime(iw.localSnap.InstallDate))
+}
+
+func (iw *infoWriter) maybePrintChinfo() {
+	if iw.diskSnap != nil {
+		return
+	}
+	chInfos := channelInfos{
+		chantpl:     "%s%s:\t%s %s%*s %*s %s\n",
+		releasedfmt: "2006-01-02",
+		esc:         iw.esc,
+	}
+	if iw.absTime {
+		chInfos.releasedfmt = time.RFC3339
+	}
+	if iw.remoteSnap != nil && iw.remoteSnap.Channels != nil && iw.remoteSnap.Tracks != nil {
+		iw.Flush()
+		chInfos.chantpl = "%s%s:\t%s\t%s\t%*s\t%*s\t%s\n"
+		chInfos.addFromRemote(iw.remoteSnap)
+	}
+	if iw.localSnap != nil {
+		chInfos.addFromLocal(iw.localSnap)
+	}
+	chInfos.dump(iw)
+}
+
+func (iw *infoWriter) maybePrintBase() {
+	if iw.verbose && iw.theSnap.Base != "" {
+		fmt.Fprintf(iw, "base:\t%s\n", iw.theSnap.Base)
+	}
+}
+
+func (iw *infoWriter) maybePrintPath() {
+	if iw.diskSnap == nil {
+		return
+	}
+	// is diskSnap -> arg is the path
+	fmt.Fprintf(iw, "path:\t%q\n", iw.arg)
+}
+
+func (iw *infoWriter) printName() {
+	fmt.Fprintf(iw, "name:\t%s\n", iw.theSnap.Name)
+}
+
+func (iw *infoWriter) printSummary() {
+	// simplest way of checking to see if it needs quoting is to try
+	raw := strings.TrimSpace(iw.theSnap.Summary)
+	type T struct {
+		S string
+	}
+	if len(raw) == 0 {
+		raw = `""`
+	} else if err := yaml.UnmarshalStrict([]byte("s: "+raw), &T{}); err != nil {
+		raw = strconv.Quote(raw)
+	}
+
+	wrapFlow(iw, []rune(raw), "summary:\t", iw.termWidth)
+}
+
+func (iw *infoWriter) maybePrintPublisher() {
+	if iw.diskSnap != nil {
+		// snaps read from disk won't have a publisher
+		return
+	}
+	fmt.Fprintf(iw, "publisher:\t%s\n", longPublisher(iw.esc, iw.theSnap.Publisher))
+}
+
+func (iw *infoWriter) maybePrintVersion() {
```

Comment:
should we call it maybePrintStandaloneVersion because is only for the disk snap case?  or not worth?